src/blob_file_builder.cc

```diff
@@ -39,11 +59,81 @@ void BlobFileBuilder::Add(const BlobRecord& record, BlobHandle* handle) {
   }
 }
 
+void BlobFileBuilder::EnterUnbuffered() {
+  // Using collected samples to train the compression dictionary
+  // Then replay those records in memory, encode them to blob file
+  // When above things are done, transform builder state into unbuffered
+  std::string samples = "";
+  std::vector<size_t> sample_lens;
+
+  for (const auto& rec_str : sample_records_) {
+    size_t copy_len =
+        cf_options_.blob_file_compression_options.zstd_max_train_bytes > 0
+            ? std::min(cf_options_.blob_file_compression_options
+                               .zstd_max_train_bytes -
+                           rec_str.size(),
+                       rec_str.size())
+            : rec_str.size();
+    samples.append(rec_str, 0, copy_len);
+    sample_lens.emplace_back(copy_len);
+  }
+  std::string dict;
+  dict = ZSTD_TrainDictionary(
+      samples, sample_lens,
+      cf_options_.blob_file_compression_options.max_dict_bytes);
+
+  compression_dict_.reset(
+      new CompressionDict(dict, cf_options_.blob_file_compression,
+                          cf_options_.blob_file_compression_options.level));
+  encoder_.SetCompressionDict(*compression_dict_);
+
+  builder_state_ = BuilderState::kUnbuffered;
+
+  // add history buffer
+  BlobHandle handle;
+  for (const std::string& rec_str : sample_records_) {
+    BlobRecord rec;
+    Slice rec_slice(rec_str);
+    rec.DecodeFrom(&rec_slice);
+    Add(rec, &handle);
```

Comment:
recursive calling `Add()` is convoluted. how about extract the file write logic into a method, and call it instead.